### PR TITLE
Added static migration proposal store

### PR DESCRIPTION
**Pull request by cwh42**

### CODE REVIEW

File: src/lib/migration/proposal_store.rb [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+#
+# To contact Novell about this file by physical or electronic mail, you may find
+# current contact information at www.novell.com.
+# ------------------------------------------------------------------------------
+
+require "yast"
+
+module Migration
+  # 1. Provides access to metadata of proposal parts (clients), as defined in the control file elements
+  # /productDefines/proposals/proposal: https://github.com/yast/yast-installation-control/blob/master/control/control.rnc
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: it is no longer valid

File: src/lib/migration/proposal_store.rb [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+  class ProposalStore < Installation::ProposalStore
+    include Yast::Logger
+    include Yast::I18n
+
+    # @ param[String] proposal_mode one of initial, service, network, hardware,
+    #   uml, ... or anything else
+    def initialize(proposal_mode)
+      textdomain "installation"
+
+      @proposal_mode = proposal_mode
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: proposal mode is not needed only to be passed to below class. so simple `super(proposal_mode)` is enough

File: src/lib/migration/proposal_store.rb [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+      @proposal_headline = _("Migration proposal")
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: not needed to store it in variable. Just put it in headline method.

File: src/lib/migration/proposal_store.rb [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+      @proposal_text_domain = "textdomain"
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: can be removed

File: src/lib/migration/proposal_store.rb [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+      @proposal_names = [ "one", "two", "three" ]
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: can be constant

File: src/lib/migration/proposal_store.rb [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+      @proposal_properties = { "key" => "value" }
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: also should be constant

File: src/lib/migration/proposal_store.rb [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+      @modules_order = []
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: not needed for now. I think we can have same order and presentation order and if it is needed in future we can add it.

File: src/lib/migration/proposal_store.rb [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    def proposal_names
+      @proposal_names
+    end
+
+    # returns single list of modules presentation order or list of tabs with list of modules
+    def presentation_order
+      return @modules_order if @modules_order
+
+      tabs? ? order_with_tabs : order_without_tabs
+    end
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: simply just return constant

File: src/lib/migration/proposal_store.rb [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+        end
+
+        if proposal_map["language_changed"]
+          @descriptions = nil # invalid descriptions cache
+          return make_proposals(force_reset: force_reset, language_changed: true)
+        end
+
+        break if proposal_map["warning_level"] == :fatal
+      end
+    end
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: remove, original implementation is good so do not repeat it

File: src/lib/migration/proposal_store.rb [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+          description["id"] = "module_#{missing_no}"
+          missing_no += 1
+        end
+
+        @id_mapping[description["id"]] = client
+
+        res[client] = description
+      end
+    end
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: same here

File: src/lib/migration/proposal_store.rb [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+
+    # @return [String] an id provided by the description API
+    def id_for(client)
+      descriptions[client]["id"]
+    end
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: same here

File: src/lib/migration/proposal_store.rb [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+
+    def title_for(client)
+      descriptions[client]["rich_text_title"] ||
+        descriptions[client]["rich_text_raw_title"] ||
+        client
+    end
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: same here

File: src/lib/migration/proposal_store.rb [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+        return nil
+      end
+
+      data = {
+        "has_next"  => false,
+        "chosen_id" => link
+      }
+
+      Yast::WFM.CallFunction(client, ["AskUser", data])
+    end
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: same here

File: src/lib/migration/proposal_store.rb [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, contact Novell, Inc.
+#
+# To contact Novell about this file by physical or electronic mail, you may find
+# current contact information at www.novell.com.
+# ------------------------------------------------------------------------------
+
+require "yast"
+
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: missing require for "installation/proposal_order". Also please add to spec file dependency on yast2-installation

File: src/lib/migration/proposal_store.rb [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+          # Generic help text for other proposals (not basic installation or
+          # hardhware configuration.
+          # General part ("You can change values...") is added as the next paragraph.
+          _(
+            "<p>\n" \
+              "To use the settings as displayed, press <b>Next</b>.\n" \
+              "</p>\n"
+          )
+        end
+      end
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: Shorten it and write only migration specific help like. `To start online migration, press <b>next</b>`

File: src/lib/migration/proposal_store.rb [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+      @proposal_properties
+    end
+
+    def order_without_tabs
+      @modules_order.sort_by! { |m| m[1] || 50 } # second element is presentation order
+
+      @modules_order.map!(&:first)
+
+      @modules_order
+    end
```
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: just return constant here

File: src/lib/migration/proposal_store.rb (hunk 5 above)
Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: current proposals should be `update_proposal` for now, others can be added later

File: src/lib/migration/proposal_store.rb (hunk 15 above)
Comment (cwh42) [inline src/lib/migration/proposal_store.rb]: So we do not need the other if cases?

Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: no, it is specific helps for different proposal modes ( one mode is installation one, others are usually deprecated ). Control file used to have more modes defined and proposal_mode switched it.

Comment (jreidinger) [inline src/lib/migration/proposal_store.rb]: to make it clear. No, it is not needed.

**Pull request closed without merging**